R/joins.R

```diff
@@ -31,49 +31,61 @@
 #' @name join.tbl_dt
 NULL
 
-join_dt <- function(op) {
-  # nocov start
-  template <- substitute(function(x, y, by = NULL, copy = FALSE, ...) {
-    by <- dplyr::common_by(by, x, y)
-    if (!identical(by$x, by$y)) {
-      stop("Data table joins must be on same key", call. = FALSE)
-    }
-    y <- dplyr::auto_copy(x, y, copy = copy)
-
-    x <- data.table::copy(x)
-    y <- data.table::copy(y)
-    data.table::setkeyv(x, by$x)
-    data.table::setkeyv(y, by$x)
-    out <- op
-    grouped_dt(out, groups(x))
-  })
-
-  f <- eval(template, parent.frame())
-  attr(f, "srcref") <- NULL # fix so prints correctly
-  f
-  # nocov end
+#' @rdname join.tbl_dt
+inner_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all = FALSE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
 }
 
 #' @rdname join.tbl_dt
-inner_join.data.table <- join_dt({merge(x, y, by = by$x, allow.cartesian = TRUE)})
+left_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all.x = TRUE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
+}
 
 #' @rdname join.tbl_dt
-left_join.data.table  <- join_dt({merge(x, y, by = by$x, all.x = TRUE, allow.cartesian = TRUE)})
+right_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all.y = TRUE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
+}
 
 #' @rdname join.tbl_dt
-right_join.data.table  <- join_dt(merge(x, y, by = by$x, all.y = TRUE, allow.cartesian = TRUE))
+semi_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...) {
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  by_x <- by$x
+  by_y <- by$y
+  y_filter <- y[, by_y, with = FALSE]
+  names(y_filter) <- by_x
```

Comment:
Perhaps `y_trimmed <- trim_y_for_semi_join(y, by)`, or even a better verb than "trim", instead of `y_filter`? This removes duplication and adds readability IMO.